src/java/org/apache/cassandra/service/accord/async/AsyncOperation.java

```diff
@@ -30,14 +30,17 @@
 import accord.local.CommandStore;
 import accord.local.PreLoadContext;
 import accord.local.SafeCommandStore;
+import accord.primitives.RoutableKey;
 import accord.primitives.Seekables;
 import accord.primitives.TxnId;
+import accord.utils.Invariants;
+import accord.utils.async.AsyncChains;
 import org.apache.cassandra.service.accord.AccordCommandStore;
-import org.apache.cassandra.service.accord.AccordCommandStore.SafeAccordCommandStore;
-import org.apache.cassandra.service.accord.api.PartitionKey;
-import org.apache.cassandra.utils.concurrent.AsyncPromise;
+import org.apache.cassandra.service.accord.AccordLiveCommand;
+import org.apache.cassandra.service.accord.AccordLiveCommandsForKey;
+import org.apache.cassandra.service.accord.AccordSafeCommandStore;
 
-public abstract class AsyncOperation<R> extends AsyncPromise<R> implements Runnable, Function<SafeCommandStore, R>, BiConsumer<Object, Throwable>
+public abstract class AsyncOperation<R> extends AsyncChains.Head<R> implements Runnable, Function<SafeCommandStore, R>
```

Comment:
@aweisberg pointed out in slack that there is an issue when listeners are `isTransient() == true`, if we drop from the cache we loose the listener!  My feature branch I added a test case to show this.  This means that `ReadData` and `WaitOnCommit` can get lost and never called

Reply:
yes that's true and more or less by design.... I don't think we should keep them around forever. If we want to change their behavior though, I don't think it should be done as part of this ticket.

Reply:
>  I don't think we should keep them around forever.

I agree it shouldn't be forever, and is not expected to work on reboot... but should a random cache evection cause such listeners to drop?  This is a hard semantic to work with, so think we need some solution for this.

> I don't think it should be done as part of this ticket.

Given this is a bug in the cep-15-accord branch, this branch not fixing this bug is fine.  I do think we should have a failing test to highlight this issue though; feel we should not go to trunk without fleshing this out

Reply:
The idea was that transient listeners support client requests, so the worst outcome of an eviction is a timed out request. I could see putting a minimum time bound on cache eviction that is >= request timeout. If we need finer grained control over this though, we should move it to another ticket / PR

Reply:
That's not quite true, transient listeners may be used when applying transactions that outlive a client query. But I think on recovery we always send enough information with Commit, so it would only be the initial request that would fail - but this would have knock-on effects. I think we really want to maintain our transient listener collection separately. But I agree we can do this in a separate PR.